src/backend/distributed/planner/multi_planner.c

```diff
@@ -117,38 +187,154 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 
 
 /*
- * CreatePhysicalPlan encapsulates the logic needed to transform a particular
- * query into a physical plan. For modifications, queries immediately enter
- * the physical planning stage, since they are essentially "routed" to remote
- * target shards. SELECT queries go through the full logical plan/optimize/
- * physical plan process needed to produce distributed query plans.
+ * IsModifyCommand returns true if the query performs modifications, false
+ * otherwise.
  */
-MultiPlan *
-CreatePhysicalPlan(Query *originalQuery, Query *query,
-				   RelationRestrictionContext *restrictionContext)
+bool
+IsModifyCommand(Query *query)
 {
-	MultiPlan *physicalPlan = MultiRouterPlanCreate(originalQuery, query,
-													restrictionContext);
-	if (physicalPlan == NULL)
+	CmdType commandType = query->commandType;
+
+	if (commandType == CMD_INSERT || commandType == CMD_UPDATE ||
+		commandType == CMD_DELETE || query->hasModifyingCTE)
 	{
-		/* Create and optimize logical plan */
-		MultiTreeRoot *logicalPlan = MultiLogicalPlanCreate(query);
-		MultiLogicalPlanOptimize(logicalPlan);
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * CreateDistributedPlan creates the distributed plan for a query over
+ * distributed tables.
+ */
+static PlannedStmt *
+CreateDistributedPlan(PlannedStmt *localPlan, Query *originalQuery, Query *query,
+					  ParamListInfo boundParams,
+					  RelationRestrictionContext *restrictionContext)
+{
+	MultiPlan *physicalPlan = NULL;
+	PlannedStmt *resultPlan = NULL;
+	DeferredErrorMessage *error = NULL;
+	bool hasUnresolved = false;
```

Comment:
Alternatively, I've been thinking whether we should call this something like "suppressPlan" (or introduce a second variable with the same value) to make it clearer what happens when this is true.